File: frontend/src/components/benchmark/content/resource_allocation/implementations/resource_allocation.cpp
#include <vector>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>

using namespace std;

class ResourceAllocation {
private:
    unordered_map<int, unordered_set<int>> adjacency;
    unordered_set<int> nodes;

public:
    void add_edge(int u, int v) {
        adjacency[u].insert(v);
        adjacency[v].insert(u);
        nodes.insert(u);
        nodes.insert(v);
    }

    void build_from_edges(const vector<pair<int, int>>& edges) {
        for (const auto& edge : edges) {
            add_edge(edge.first, edge.second);
        }
    }

    double resource_allocation_score(int u, int v) {
        if (adjacency.find(u) == adjacency.end() || adjacency.find(v) == adjacency.end()) {
            return 0.0;
        }

        double score = 0.0;
        for (int node : adjacency[u]) {
            if (adjacency[v].find(node) != adjacency[v].end()) {
                int degree = adjacency[node].size();
                if (degree > 0) {
                    score += 1.0 / degree;
                }
            }
        }
        return score;
    }

    vector<pair<pair<int, int>, double>> score_all_pairs() {
        vector<pair<pair<int, int>, double>> scores;
        vector<int> node_list(nodes.begin(), nodes.end());

        for (size_t i = 0; i < node_list.size(); i++) {
            for (size_t j = i + 1; j < node_list.size(); j++) {
                int u = node_list[i];
                int v = node_list[j];
                double score = resource_allocation_score(u, v);
                if (score > 0) {
                    scores.push_back({{u, v}, score});
                }
            }
        }

        sort(scores.begin(), scores.end(),
             [](const auto& a, const auto& b) {
                 return a.second > b.second;
             });

        return scores;
    }

    vector<pair<pair<int, int>, double>> top_k_predictions(int k = 10) {
        auto scores = score_all_pairs();
        vector<pair<pair<int, int>, double>> result;
        for (int i = 0; i < min(k, (int)scores.size()); i++) {
            result.push_back(scores[i]);
        }
        return result;
    }
};

File: frontend/src/components/benchmark/content/resource_allocation/implementations/resource_allocation.php
<?php

class ResourceAllocation {
    private array $adjacency = [];
    private array $nodes = [];

    public function addEdge(int $u, int $v): void {
        if (!isset($this->adjacency[$u])) {
            $this->adjacency[$u] = [];
        }
        if (!isset($this->adjacency[$v])) {
            $this->adjacency[$v] = [];
        }
        $this->adjacency[$u][$v] = true;
        $this->adjacency[$v][$u] = true;
        $this->nodes[$u] = true;
        $this->nodes[$v] = true;
    }

    public function buildFromEdges(array $edges): void {
        foreach ($edges as $edge) {
            $this->addEdge($edge[0], $edge[1]);
        }
    }

    public function resourceAllocationScore(int $u, int $v): float {
        if (!isset($this->adjacency[$u]) || !isset($this->adjacency[$v])) {
            return 0.0;
        }

        $score = 0.0;
        foreach (array_keys($this->adjacency[$u]) as $node) {
            if (isset($this->adjacency[$v][$node])) {
                $degree = count($this->adjacency[$node]);
                if ($degree > 0) {
                    $score += 1.0 / $degree;
                }
            }
        }
        return $score;
    }

    public function scoreAllPairs(): array {
        $scores = [];
        $nodeList = array_keys($this->nodes);

        for ($i = 0; $i < count($nodeList); $i++) {
            for ($j = $i + 1; $j < count($nodeList); $j++) {
                $u = $nodeList[$i];
                $v = $nodeList[$j];
                $score = $this->resourceAllocationScore($u, $v);
                if ($score > 0) {
                    $scores[] = [
                        'pair' => [$u, $v],
                        'score' => $score
                    ];
                }
            }
        }

        usort($scores, function($a, $b) {
            return $b['score'] <=> $a['score'];
        });

        return $scores;
    }

    public function topKPredictions(int $k = 10): array {
        $scores = $this->scoreAllPairs();
        return array_slice($scores, 0, $k);
    }
}

?>